File: algoexpert/easy/09.product_sum.c
#include <stdio.h>

typedef struct List {
  int value;
  struct List *head;
  struct List *tail;
} List;

// FIXME: outputs 23, when it should output 71
int sum_product_aux(struct List *list, int depth) {
  int sum = 0;

  if (list->head)
    sum += sum_product_aux(list->head, depth + 1);
  else
    sum += list->value;

  if (list->tail)
    sum += sum_product_aux(list->tail, depth);

  return sum * depth;
}

int sum_product(struct List *list) { return sum_product_aux(list, 1); }

void foo(struct List list) {}

int main(void) {
  struct List list = {
      1,
      NULL,
      &(List){
          2,
          NULL,
          &(List){3, NULL,
                  &(List){0,
                          &(List){1, NULL,
                                  &(List){
                                      3,
                                      NULL,
                                      NULL,
                                  }},
                          &(List){
                              3,
                              NULL,
                              NULL,
                          }}},
      },
  };
  printf("sum product is: %d\n", sum_product(&list));
  return 0;
}

File: algoexpert/easy/09.product_sum.go
package main

import "fmt"

type arr struct {
	value int
	curr  *arr
	next  *arr
}

// FIXME: outputs 23, when it should output 17
func solve(input arr, depth int) int {
	var sum int

	if input.curr != nil {
		sum += solve(*input.curr, depth+1)
	} else {
		sum += input.value
	}

	if input.next != nil {
		sum += solve(*input.next, depth)
	}

	return sum * depth
}

func main() {
	input := arr{1, nil, &arr{2, nil, &arr{3, nil, &arr{
		0,
		&arr{1, nil, &arr{3, nil, nil}},
		&arr{3, nil, nil}}}},
	}
	output := solve(input, 1)
	fmt.Println(output)
}
